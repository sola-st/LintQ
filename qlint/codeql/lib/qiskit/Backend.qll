import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Register
import qiskit.Gate
import qiskit.Circuit

/** Simulator backend. */
class Backend extends DataFlow::CallCfgNode {
  // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
  Backend() {
    this = API::moduleImport("qiskit").getMember("Aer").getMember("get_backend").getACall() or
    this = API::moduleImport("qiskit").getMember("BasicAer").getMember("get_backend").getACall()
  }

  /** The name of the backend (according to the initialization string). */
  string getKind() {
    // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
    // > 'qasm_simulator'
    exists(StrConst str, DataFlow::LocalSourceNode strSource |
      strSource = this.getArg(0) and
      strSource.asExpr() = str
    |
      result = str.getText()
    )
  }

  /** If the simulator is statevector simulator. */
  predicate isStatevectorSimulator() {
    // case: from qiskit import Aer; Aer.get_backend('statevector_simulator')
    this.getKind().matches("%statevector%")
  }

  /** If the simulator is Unitary simulator. */
  predicate isUnitarySimulator() {
    // case: from qiskit import Aer; Aer.get_backend('unitary_simulator')
    this.getKind().matches("%unitary%")
  }

  /** Get a circuit run with this backend. */
  QuantumCircuit getACircuitToBeRun() {
    // case: Aer.get_backend('qasm_simulator').run(circuit)
    // > circuit
    exists(QuantumCircuit circuit, DataFlow::CallCfgNode runCall |
      runCall = this.getAnAttributeRead("run").getACall() and
      (
        circuit.flowsTo(runCall.getArg(0)) or
        circuit.flowsTo(runCall.getArgByName("circuits"))
      )
    |
      result = circuit
    )
    or
    // case: execute(qc, Aer.get_backend('qasm_simulator'), shots=1)
    // > qc
    exists(QuantumCircuit circuit, ExecuteCall executeCall |
      (
        circuit.flowsTo(executeCall.getArg(0)) or
        circuit.flowsTo(executeCall.getArgByName("experiments"))
      ) and
      (
        this.flowsTo(executeCall.getArg(1)) or
        this.flowsTo(executeCall.getArgByName("backend"))
      )
    |
      result = circuit
    )
  }

  /** Get a backend run with this backend. */
  BackendRun getABackendRun() {
    exists(BackendRun bkdRun |
      bkdRun.getBackend() = this
    |
      result = bkdRun
    )
  }

}

/** Backend run.
 *
 * This is a call to a backend run method.
 */
abstract class BackendRun extends DataFlow::CallCfgNode {

  abstract Backend getBackend();

}

/** A call to a backend run method. */
class BackendRunViaRunCall extends BackendRun {
  BackendRunViaRunCall() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      this = runCall
    )
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      result = backend
    )
  }
}

/** A call to a backend execute method. */
class BackendRunViaExecuteCall extends BackendRun {
  BackendRunViaExecuteCall() {
    this instanceof ExecuteCall
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend |
      (
        backend.flowsTo(this.getArg(1)) or
        backend.flowsTo(this.getArgByName("backend"))
      )
    |
      result = backend
    )
  }
}

/** Result of a simulator run. */
class BackendResult extends DataFlow::CallCfgNode {
  BackendResult() {
    exists(BackendRun bkdRun |
      this = bkdRun.getAnAttributeRead("result").getACall()
      or
      this = bkdRun.getAnAttributeReference("result")
      or
      this = bkdRun.getAnAttributeWrite("result")
    )
  }

  /** The backend run that generated this result. */
  BackendRun getBackendRun() {
    exists(BackendRun bkdRun |
      // connect the run and its result
      this = bkdRun.getAnAttributeRead("result").getACall() or
      this = bkdRun.getAnAttributeReference("result") or
      this = bkdRun.getAnAttributeWrite("result")
    |
      result = bkdRun
    )
  }

}


/** Statevector returned by a result of a simulator run. */
class Statevector extends DataFlow::CallCfgNode {
  Statevector() {
    exists(BackendResult bkdResult |
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    )
  }

  /** The backend run that generated this statevector. */
  BackendRun producedByBackendRun() {
    exists(BackendResult bkdResult, BackendRun bkdRun |
      // connect the run and its result
      bkdResult.getBackendRun() = bkdRun and
      // the result has been called to extract the statevector
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    |
      result = bkdRun
    )
  }

}