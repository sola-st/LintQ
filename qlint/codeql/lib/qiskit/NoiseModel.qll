
import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs


class PauliError extends DataFlow::CallCfgNode {

    PauliError() {
        // OLDER API
        // from qiskit.providers.aer.noise.errors import pauli_error
        this = API::moduleImport("qiskit").getMember("providers")
            .getMember("aer").getMember("noise").getMember("errors")
            .getMember("pauli_error").getACall()
        or
        // NEWER API
        this = API::moduleImport("qiskit_aer").getMember("noise")
            .getMember("pauli_error").getACall()
    }

    StrConst getPauliString() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        exists(
            StrConst strConst
            |
            this.getArg(0).asCfgNode().getNode().contains(strConst)
            |
            result = strConst
            )
    }

    /** Check that the Pauli String are of different size.*/
    predicate arePauliStringSameSize() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        not exists(
            string str1, string str2
            |
            this.getPauliString().getText() = str1
            and
            this.getPauliString().getText() = str2
            |
            str1.length() != str2.length()
            )
    }


}
