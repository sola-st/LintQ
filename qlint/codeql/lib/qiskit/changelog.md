2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))