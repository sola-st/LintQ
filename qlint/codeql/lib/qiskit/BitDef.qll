import qiskit.Circuit
import qiskit.Register
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs

/** Definition of a qubit or bit. */
abstract class BitDefinition extends DataFlow::LocalSourceNode {
  int getAnIndex() {
    if exists(int i | i = this.getAnIndexIfAny())
    then exists(int i | i = this.getAnIndexIfAny() | result = i)
    else result = -1
  }

  string getARegisterName() {
    if exists(RegisterV2 reg | reg = this.getARegister())
    then exists(RegisterV2 reg | reg = this.getARegister() | result = reg.getName())
    else result = "anonymous register"
  }

  string getACircuitName() {
    if exists(QuantumCircuit circ | circ = this.getACircuit())
    then exists(QuantumCircuit circ | circ = this.getACircuit() | result = circ.getName())
    else result = "anonymous circuit"
  }

  abstract string getTypeName();

  abstract int getAnIndexIfAny();

  abstract RegisterV2 getARegister();

  abstract QuantumCircuit getACircuit();

  boolean equals(BitDefinition other) {
    if
      this.getARegisterName() = other.getARegisterName() and
      this.getAnIndexIfAny() = other.getAnIndexIfAny() and
      this.getACircuitName() = other.getACircuitName() and
      this.getARegister() = other.getARegister() and
      this.getACircuit() = other.getACircuit() and
      this.getTypeName() = other.getTypeName()
    then result = true
    else result = false
  }
}

abstract class QubitDefinition extends BitDefinition {
  override string getTypeName() { result = "qubit" }
}

abstract class ClbitDefinition extends BitDefinition {
  override string getTypeName() { result = "clbit" }
}

predicate isIntegerParameterOfCall(
  DataFlow::LocalSourceNode integerParameter, DataFlow::CallCfgNode call, int indexParam
) {
  exists(IntegerLiteral intLit |
    integerParameter.flowsTo(call.getArg(indexParam)) and
    integerParameter.asExpr() = intLit
  )
}

/** Qubit generated implicitly by QuantumCircuit(). */
class ImplicitCircuitQubitDefinition extends BitDefinition, QubitDefinition {
  ImplicitCircuitQubitDefinition() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0)
    )
  }

  /** Returns an index of the qubit in the register. */
  override int getAnIndexIfAny() {
    // qc = QuantumCircuit(4, 3)
    // >> [0, 1, 2, 3]
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource, int i |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      i in [0 .. qc.getNumberOfQubits() - 1] and
      result = i
    )
  }

  override RegisterV2 getARegister() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      result = qc
    )
  }

  override QuantumCircuit getACircuit() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      result = qc
    )
  }
}

/** Clbit generated implicitly by QuantumCircuit(). */
class ImplicitCircuitClbitDefinition extends BitDefinition, ClbitDefinition {
  ImplicitCircuitClbitDefinition() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1)
    )
  }

  /** Returns an index of the clbit in the register. */
  override int getAnIndexIfAny() {
    // qc = QuantumCircuit(4, 3)
    // >> [0, 1, 2]
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource, int i |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      i in [0 .. qc.getNumberOfClassicalBits() - 1] and
      result = i
    )
  }

  override RegisterV2 getARegister() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      result = qc
    )
  }

  override QuantumCircuit getACircuit() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      result = qc
    )
  }
}

/** Bits generated implicitly by QuantumRegister() and ClassicalRegister(). */
abstract class ImplicitRegisterBitDefinition extends BitDefinition {
  /** Returns an index of the bit in the register. */
  override int getAnIndexIfAny() {
    exists(int i | i in [0 .. this.(RegisterV2).getSize() - 1] | result = i)
  }

  /** Returns the register of the bit. */
  override RegisterV2 getARegister() { result = this }

  /** Returns the circuit in which the bit is added. */
  override QuantumCircuit getACircuit() {
    // get the circuit from the register
    result = this.getARegister().getACircuit()
  }
}

/** Qubit generated implicitly by QuantumRegister(). */
class ImplicitRegisterQubitDefinition extends ImplicitRegisterBitDefinition, QubitDefinition {
  ImplicitRegisterQubitDefinition() {
    exists(QuantumRegisterV2 qreg, int i | i in [0 .. qreg.getSize() - 1] | this = qreg)
  }
}

/** Clbit generated implicity by ClassicalRegister() */
class ImplicitRegisterClbitDefinition extends ImplicitRegisterBitDefinition, ClbitDefinition {
  ImplicitRegisterClbitDefinition() {
    exists(ClassicalRegisterV2 creg, int i | i in [0 .. creg.getSize() - 1] | this = creg)
  }
}

// EXPLICIT INITIALIZATION OF SINGLE BITS
/** Bits generated by Qubit() or Clbit(). */
abstract class ExplicitSingleBitDefinition extends BitDefinition {
  /** Returns the name of the identifier of the bit. */
  string getName() {
    exists(AssignStmt a |
      a.contains(this.(DataFlow::CallCfgNode).getNode().getNode()) and
      result = a.getATarget().(Name).getId()
    )
  }

  /** Return the index paramer of the current bit. */
  override int getAnIndexIfAny() {
    // single_bit = Clbit(register=creg, index=2)
    result =
      this.(API::CallNode)
          .getParameter(1, "index")
          .getAValueReachingSink()
          .asExpr()
          .(IntegerLiteral)
          .getValue()
  }

  /** Returns the register of the current bit. */
  override RegisterV2 getARegister() {
    // single_bit = Clbit(register=creg, index=2)
    exists(RegisterV2 reg |
      reg.asExpr() =
        this.(API::CallNode).getParameter(0, "register").getAValueReachingSink().asExpr()
    |
      result = reg
    )
  }

  /** Returns the circuit in which the bit is added. */
  override QuantumCircuit getACircuit() {
    // get the circuit from the register
    result = this.getARegister().getACircuit()
  }
}

/** Qubit generated by Qubit(). */
class ExplicitSingleQubitDefinition extends ExplicitSingleBitDefinition, QubitDefinition {
  ExplicitSingleQubitDefinition() {
    this = API::moduleImport("qiskit").getMember("Qubit").getACall()
    or
    this = API::moduleImport("qiskit").getMember("circuit").getMember("Qubit").getACall()
  }
}

/** Clbit generated by Clbit(). */
class ExplicitSingleClbitDefinition extends ExplicitSingleBitDefinition, ClbitDefinition {
  ExplicitSingleClbitDefinition() {
    this = API::moduleImport("qiskit").getMember("Clbit").getACall()
    or
    this = API::moduleImport("qiskit").getMember("circuit").getMember("Clbit").getACall()
  }
}
