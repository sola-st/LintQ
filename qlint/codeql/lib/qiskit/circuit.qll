import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}