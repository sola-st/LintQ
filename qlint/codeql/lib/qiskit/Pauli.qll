import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.NoiseModel


class PauliString extends StrConst {
    PauliString() {
        // it can be used in a PauliError
        exists(
            PauliError pauliError
            |
            pauliError.getPauliString() = this
        )
        or
        // it can be used in a Pauli object
        exists(
            DataFlow::CallCfgNode pauliObject
            |
            pauliObject = API::moduleImport("qiskit").getMember("circuit")
                        .getMember("QuantumCircuit").getMember("pauli").getACall()
            or
            // from qiskit.quantum_info import Pauli
            pauliObject = API::moduleImport("qiskit").getMember("quantum_info")
                        .getMember("Pauli").getACall()
            |
            this = pauliObject.getArg(0).asCfgNode().getNode()
        )
    }

    /** Check if the Pauli satisfy the regex ^[+-]?1?[ij]?[IXYZ]+$ */
    predicate isValid() {
        this.getText().regexpMatch("^[+-]?1?[ij]?[IXYZ]+$")
    }

}