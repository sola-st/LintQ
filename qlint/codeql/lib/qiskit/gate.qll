import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.circuit

class GateName extends string {
    GateName() {
        this = "x" or
        this = "y" or
        this = "z" or
        this = "h" or
        this = "s" or
        this = "sdg" or
        this = "t" or
        this = "tdg" or
        this = "rx" or
        this = "ry" or
        this = "rz" or
        this = "u1" or
        this = "u2" or
        this = "u3" or
        this = "cx" or
        this = "cy" or
        this = "cz" or
        this = "ch" or
        this = "crz" or
        this = "cu1" or
        this = "cu3" or
        this = "swap" or
        this = "ccx" or
        this = "cswap" or
        this = "rxx" or
        this = "ryy" or
        this = "rzz" or
        this = "rzx" or
        this = "rzz" or
        this = "measure" or
        this = "measure_all"
    }
}

// TODO: improve to support
// from qiskit import QuantumCircuit
// qc = QuantumCircuit(2)
// qc.append(CXGate(), [0, 1])

class GenericGate extends DataFlow::CallCfgNode {

    GenericGate() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name|
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }

    string get_gate_name() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall() |
            result = a_supported_gate_name
        )
    }

    QuantumCircuit get_quantum_circuit() {
        exists(
            QuantumCircuit circ |
            this = circ.getAnAttributeRead(this.get_gate_name()).getACall()|
            result = circ
        )
    }


    int get_a_target_qubit() {
        // TODO: improve to support sequences of qubits or integers
        exists(
            QuantumCircuit circ, int target_qubit, int i |
            this = circ.getAnAttributeRead().getACall() and
            target_qubit = this.getArg(i).asExpr().(IntegerLiteral).getValue()|
            // return a list with only the target qubit
            result = target_qubit
        )
        // handle the case where the arguments are in accessed through a quantum
        // reigster variable like this:
        // q = QuantumRegister(2)
        // qc = QuantumCircuit(q)
        // qc.cx(q[0], q[1])
        or
        exists(
            QuantumCircuit circ,
            QuantumRegister qreg,
            DataFlow::Node nd,
            DataFlow::ExprNode targetSubscript,
            Subscript subscript,
            IntegerLiteral bit,
            int i |
                this = circ.getAnAttributeRead().getACall() and
                qreg.flowsTo(nd) and
                nd.asExpr() = targetSubscript.asExpr() and
                targetSubscript.asExpr() = subscript.getObject() and
                subscript = this.getArg(i).asExpr() and
                bit = subscript.getIndex() |
                result = bit.getValue()
        )
    }

    QuantumRegister get_a_target_qubit_in_register() {
        exists(
            QuantumCircuit circ,
            QuantumRegister qntReg |
                this = circ.getAnAttributeRead().getACall() and
                qntReg = circ.get_a_quantum_register() and
                qntReg.flowsTo(this.getArg(0)) |
                result = qntReg
        )
    }

    predicate follows(GenericGate g) {
        exists(
            QuantumCircuit circ |
                this = circ.getAnAttributeRead().getACall() and
                g = circ.getAnAttributeRead().getACall() and
                not this = g and
                this.get_a_target_qubit() = g.get_a_target_qubit() and
                this.get_quantum_circuit() = g.get_quantum_circuit() and
                g.asCfgNode().strictlyReaches(this.asCfgNode())
        )
    }
}


class Gate extends GenericGate {
    Gate() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name != "measure"  and a_supported_gate_name != "measure_all"|
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }
}

class Measure extends GenericGate {
    Measure() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name = "measure" |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }

    override int get_a_target_qubit() {
        // note that measure gets qubits only in the first position
        // TODO: improve to support sequences of qubits or integers
        // e.g. qc.measure([0, 1], [0, 1])
        // from documentation Parameters
        // qubit (Union[Qubit, QuantumRegister, int, slice, Sequence[Union[Qubit, int]]]) – qubit to measure.
        // cbit (Union[Clbit, ClassicalRegister, int, slice, Sequence[Union[Clbit, int]]]) – classical bit to place the measurement in.

        // TODO: improve to support entire registers
        // e.g. qc.measure(q, c)
        exists(
            QuantumCircuit circ, int target_qubit|
            this = circ.getAnAttributeRead("measure").getACall() and
            target_qubit = this.getArg(0).asExpr().(IntegerLiteral).getValue()|
            // return a list with only the target qubit
            result = target_qubit
        )
        or
        exists(
            QuantumCircuit circ,
            QuantumRegister qreg,
            DataFlow::Node nd,
            DataFlow::ExprNode targetSubscript,
            Subscript subscript,
            IntegerLiteral bit|
                this = circ.getAnAttributeRead("measure").getACall() and
                qreg.flowsTo(nd) and
                nd.asExpr() = targetSubscript.asExpr() and
                targetSubscript.asExpr() = subscript.getObject() and
                subscript = this.getArg(0).asExpr() and
                bit = subscript.getIndex() |
                result = bit.getValue()
        )
    }
}

class MeasureAll extends GenericGate {
    MeasureAll() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name = "measure_all" |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }
}