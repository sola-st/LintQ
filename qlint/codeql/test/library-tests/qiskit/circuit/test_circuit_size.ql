import python
import qiskit.circuit

from QuantumCircuit circ
where not circ.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%")
select circ, "Classical register '" + circ.get_name() + "' with " +
    circ.get_total_num_qubits() + " bits and " +
    circ.get_total_num_bits() + " qubits"
