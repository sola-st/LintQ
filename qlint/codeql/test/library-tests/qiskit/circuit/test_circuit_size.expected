| concept_circuit_parametrizable.py:9:6:9:22 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc' with 0 bits and 1 qubits |
| concept_circuit_parametrizable.py:20:10:20:26 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc' with 0 bits and 2 qubits |
| concept_circuit_parametrizable.py:27:6:27:23 | ControlFlowNode for unknown_api_call() | Quantum Circuit 'qc' with 0 bits and 0 qubits |
| concept_circuit_parametrizable.py:28:12:28:38 | ControlFlowNode for Attribute() | Quantum Circuit 'qc_bound' with 0 bits and 0 qubits |
| concept_circuit_parametrizable.py:32:10:32:32 | ControlFlowNode for EfficientSU2() | Quantum Circuit 'qc_su2' with 0 bits and 3 qubits |
| concept_circuit_parametrizable.py:35:11:35:35 | ControlFlowNode for RealAmplitudes() | Quantum Circuit 'qc_real' with 0 bits and 3 qubits |
| concept_circuit_parametrizable.py:38:9:38:46 | ControlFlowNode for TwoLocal() | Quantum Circuit 'qc_tl' with 0 bits and 3 qubits |
| concept_circuit_subcircuit.py:6:17:6:42 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_subcircuit' with 3 bits and 3 qubits |
| concept_circuit_subcircuit.py:11:12:11:27 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_macro' with 8 bits and 10 qubits |
| concept_circuit_subcircuit.py:17:12:17:26 | ControlFlowNode for Attribute() | Quantum Circuit 'qc_clone' with 8 bits and 10 qubits |
| concept_circuit_subcircuit.py:20:15:20:65 | ControlFlowNode for Attribute() | Quantum Circuit 'qc_composed' with 0 bits and 0 qubits |
| concept_circuit_subcircuit.py:32:10:32:29 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc' with 0 bits and 3 qubits |
| concept_circuit_subcircuit.py:44:10:44:29 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc' with 0 bits and 3 qubits |
| concept_circuit_subcircuit.py:48:17:48:36 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_second' with 0 bits and 3 qubits |
| concept_circuit_subcircuit.py:53:32:53:50 | ControlFlowNode for create_subcircuit() | Quantum Circuit 'qc_subcircuit_via_subroutine' with 0 bits and 0 qubits |
| concept_circuit_subcircuit.py:56:29:56:58 | ControlFlowNode for crete_subcircuit_via_compose() | Quantum Circuit 'qc_subcircuit_via_compose' with 0 bits and 0 qubits |
| concept_circuit_transpile.py:5:11:5:30 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'circuit' with 1 bits and 1 qubits |
| concept_circuit_transpile.py:8:22:8:50 | ControlFlowNode for transpile() | Quantum Circuit 'transpiled_circuit' with 0 bits and 0 qubits |
| concept_subcircuit_as_gate_instr.py:4:12:4:28 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_instr' with 0 bits and 3 qubits |
| concept_subcircuit_as_gate_instr.py:15:10:15:38 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_maj' with 0 bits and 3 qubits |
| concept_subcircuit_as_gate_instr.py:22:11:22:30 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'main_qc' with 3 bits and 3 qubits |
| pred_circuit_gate.py:6:6:6:31 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc' with 8 bits and 8 qubits |
| pred_circuit_size.py:5:15:5:40 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_only_reg' with 3 bits and 3 qubits |
| pred_circuit_size.py:7:20:7:39 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_only_integers' with 1 bits and 2 qubits |
| pred_circuit_size.py:11:19:11:34 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_add_register' with 8 bits and 10 qubits |
| pred_circuit_size.py:15:22:15:53 | ControlFlowNode for QuantumCircuit() | Quantum Circuit 'qc_multi_registers' with 11 bits and 13 qubits |
